judge/admin/contest.py

```diff
@@ -211,13 +216,48 @@ def make_hidden(self, request, queryset):
                                              count) % count)
     make_hidden.short_description = _('Mark contests as hidden')
 
+    def set_locked(self, request, queryset):
+        for row in queryset:
+            self.set_is_locked(row, True)
+        count = queryset.count()
+        self.message_user(request, ungettext('%d contest successfully locked.',
+                                             '%d contests successfully locked.',
+                                             count) % count)
+    set_locked.short_description = _('Lock contest submissions')
+
+    def set_unlocked(self, request, queryset):
+        for row in queryset:
+            self.set_is_locked(row, False)
+        count = queryset.count()
+        self.message_user(request, ungettext('%d contest successfully unlocked.',
+                                             '%d contests successfully unlocked.',
+                                             count) % count)
+    set_unlocked.short_description = _('Unlock contest submissions')
+
+    def set_is_locked(self, contest, is_locked):
+        with transaction.atomic():
+            contest.is_locked = is_locked
+            contest.save()
+            for row in ContestSubmission.objects.filter(participation__contest=contest):
+                submission = row.submission
+                submission.is_locked = is_locked
+                submission.save()
+
     def get_urls(self):
         return [
             url(r'^rate/all/$', self.rate_all_view, name='judge_contest_rate_all'),
             url(r'^(\d+)/rate/$', self.rate_view, name='judge_contest_rate'),
             url(r'^(\d+)/judge/(\d+)/$', self.rejudge_view, name='judge_contest_rejudge'),
+            url(r'^(\d+)/lock/$', self.lock_view, name='judge_contest_lock'),
         ] + super(ContestAdmin, self).get_urls()
 
+    def lock_view(self, request, id):
+        if not request.user.has_perm('judge.contest_lock'):
+            raise PermissionDenied()
+        contest = get_object_or_404(Contest, id=id)
+        self.set_is_locked(contest, not contest.is_locked)
```

Comment:
Not sure how I feel about the lack of idempotence with this toggle, and the fact it's a GET.

Reply:
This should at least be a post request, and might make more sense have separate lock and unlock operations.